source/AsyncRgbLedSimulationDataGenerator.cpp

```diff
@@ -15,57 +19,97 @@ AsyncRgbLedSimulationDataGenerator::~AsyncRgbLedSimulationDataGenerator()
 	{
-		CreateSerialByte();
+        WriteReset();
+
+        // six RGB-triple cascade between resets, i.e six discrete LEDs
+        // or two of the 3-LED combined drivers. We should perhaps make
+        // this adjustable
+		for (int t=0; t<6; ++t) {
+			CreateRGBWord();
+		}
 	}
 
-	*simulation_channel = &mSerialSimulationData;
+	*simulation_channel = &mLEDSimulationData;
 	return 1;
 }
 
-void AsyncRgbLedSimulationDataGenerator::CreateSerialByte()
+void AsyncRgbLedSimulationDataGenerator::CreateRGBWord()
 {
-	U32 samples_per_bit = mSimulationSampleRateHz / mSettings->mBitRate;
-
-	U8 byte = mSerialText[ mStringIndex ];
-	mStringIndex++;
-	if( mStringIndex == mSerialText.size() )
-		mStringIndex = 0;
+	const RGBValue rgb = RandomRGBValue();
+	WriteRGBTriple( rgb );
+}
 
-	//we're currenty high
-	//let's move forward a little
-	mSerialSimulationData.Advance( samples_per_bit * 10 );
+void AsyncRgbLedSimulationDataGenerator::WriteRGBTriple( const RGBValue& rgb )
+{
+	U16 values[3];
+	rgb.ConvertToControllerOrder(mSettings->GetColorLayout(), values);
 
-	mSerialSimulationData.Transition();  //low-going edge for start bit
-	mSerialSimulationData.Advance( samples_per_bit );  //add start bit time
+	for (int i=0; i<3; ++i) {
+    	WriteUIntData( values[i], mSettings->BitSize() );
+	}
+}
 
-	U8 mask = 0x1 << 7;
-	for( U32 i=0; i<8; i++ )
-	{
-		if( ( byte & mask ) != 0 )
-			mSerialSimulationData.TransitionIfNeeded( BIT_HIGH );
-		else
-			mSerialSimulationData.TransitionIfNeeded( BIT_LOW );
+void AsyncRgbLedSimulationDataGenerator::WriteReset()
+{
+    // interstitial after last data bit, before the reset goes low
+    mLEDSimulationData.Advance( mClockGenerator.AdvanceByTimeS(1e-5) );
+    mLEDSimulationData.Transition(); // go low
+    // scale by 120% since reset time is the minimum I believe
+    const double resetSec = mClockGenerator.AdvanceByTimeS(mSettings->ResetTimeNSec() * 1.2 * NSEC_TO_SEC);
+    mLEDSimulationData.Advance( resetSec );
+	mLEDSimulationData.Transition(); // go high to end the reset
+}
 
-		mSerialSimulationData.Advance( samples_per_bit );
+void AsyncRgbLedSimulationDataGenerator::WriteUIntData( U32 data, U8 bit_count )
+{
+	U32 mask =  1 << (bit_count - 1);
+	for( U32 bit=0; bit < bit_count; ++bit) {
+		WriteBit(data & mask);
 		mask = mask >> 1;
-	}
+	}	
+}
 
-	mSerialSimulationData.TransitionIfNeeded( BIT_HIGH ); //we need to end high
+void AsyncRgbLedSimulationDataGenerator::WriteBit(bool b)
+{
+    const BitState bs = b ? BIT_HIGH : BIT_LOW;
+    const double highSamples = mClockGenerator.AdvanceByTimeS( mSettings->DataTimeNSecHigh( bs ) * NSEC_TO_SEC );
+    const double lowSamples = mClockGenerator.AdvanceByTimeS( mSettings->DataTimeNSecLow( bs ) * NSEC_TO_SEC );
+    mLEDSimulationData.Advance( highSamples );
+	mLEDSimulationData.Transition(); // go low
+    mLEDSimulationData.Advance( lowSamples );
+	mLEDSimulationData.Transition(); // go high to end this bit
```

Comment:
Right now, this function assumes that a transition has just been added to the simulation data before getting called, and it leaves a transition on the channel after getting called.

Since we're eliminating the positive pulse that you're currently adding at the beginning of the reset pulse function, I propose the following:
1. The initial bit state is low.
2. The reset function does not add any transitions, it just calls advance. (the function enters low and exits low)
3. This function, Writebit, transitions immediately on call (as it does) but it does not transition again after the final low pulse, leaving the channel low. (the function enters low and exits low)

This results in the mLEDSimulationData always being low outside of calls to WriteBit and WriteReset.
As an example analyzer, this is an important concept. If there were 3 or more functions that all created events on the bus, the developer would need to keep track of what the state is of the simulation channel at the entrance to each function. Keeping it low outside of all functions makes it easier to reason about.

Reply:
My implementation here was based on the assumption that the line is always high by default, and was taken low-then-high by all the Write functions (WriteBit or WriteReset). So I think the only change here (which I will do) is to change the convention that we presume data signal is low on entry and exit rather than high.

Reply:
perfect